### ISSUE TITLE
Disable execution optimizations for invalid work

### ISSUE THREAD

**Issue by lptr** [gradle/gradle#15304]
Gradle software checks user code in several ways, e.g. if the necessary input/output
annotations are present, if there are overlapping outputs etc. When such problems occur, some
of the optimizations in Gradle software are not safe and may result in incorrect build results.
We can enforce work validity by punishing invalid work items by:

- disable caching for them (load or store),
- disable incremental execution,
- ignore the state of their inputs and outputs and run them every time,
- prevent other work from running in parallel with invalid work,
- invalidate the whole VFS before invalid work is executed.

---
cc: @gradle/execution

**Comment by lptr** [comment]
Duplicate of #15305.
